vcf/header.go

```diff
@@ -2,13 +2,286 @@ package vcf
+}
+
+// FormatHeader contains info encoded by header lines beginning in ##FORMAT.
+type FormatHeader struct {
+	Key
+	Description string
+}
+
+// ReadHeader reads and parses the header in a vcf file.
+func ReadHeader(er *fileio.EasyReader) Header {
+	var line string
+	var err error
+	var nextBytes []byte
+	var headerText []string
+	for nextBytes, err = er.Peek(1); err == nil && nextBytes[0] == '#'; nextBytes, err = er.Peek(1) {
+		line, _ = fileio.EasyNextLine(er)
+		headerText = append(headerText, line)
+	}
+	return parseHeader(headerText)
+}
+
+// newHeader allocates memory for a new vcf Header.
+func newHeader() Header {
+	var h Header
+	h.Info = make(map[string]InfoHeader)
+	h.Filter = make(map[string]FilterHeader)
+	h.Format = make(map[string]FormatHeader)
+	h.Chroms = make(map[string]chromInfo.ChromInfo)
+	h.Samples = make(map[string]int)
+	return h
+}
+
+// parseHeader parses the text in a vcf Header to a more useful format.
+func parseHeader(text []string) Header {
+	if len(text) == 0 { // vcf w/o header is valid
+		return Header{}
+	}
+
+	header := newHeader()
+	header.Text = text
+	var chromIdx int
+	for i := range text {
+		switch strings.Split(text[i][2:], "=")[0] { // pulls line label (e.g. "INFO" from "##INFO=<...>")
+		case "fileformat":
+			header.FileFormat = parseFileFormatFromHeader(text[i])
+
+		case "contig":
+			header.Chroms = parseChromsFromHeader(text[i], header.Chroms, chromIdx)
+			chromIdx++
+
+		case "INFO":
+			header.Info = parseInfoFromHeader(text[i], header.Info)
+
+		case "FILTER":
+			header.Filter = parseFilterFromHeader(text[i], header.Filter)
+
+		case "FORMAT":
+			header.Format = parseFormatFromHeader(text[i], header.Format)
+		}
+	}
+	header.Samples = parseSamplesFromHeader(text[len(text)-1])
+	return header
+}
+
+// parseFileFormatFromHeader parses a line beginning with ##fileformat to determine the vcf version.
+func parseFileFormatFromHeader(line string) string {
+	return strings.Split(line, "=")[1]
+}
+
+// parseChromsFromHeader parses a line beginning with ##contig to a map keying the chrom name to the chromInfo.
+func parseChromsFromHeader(line string, chroms map[string]chromInfo.ChromInfo, chromIdx int) map[string]chromInfo.ChromInfo {
+	fields := getHeaderFields(line)
+	var chrom chromInfo.ChromInfo
+	chrom.Order = chromIdx
+	var err error
+
+	// Parse chromInfo
+	for i := range fields {
+		switch {
+		case strings.HasPrefix(fields[i], "ID="):
+			chrom.Name = fields[i][3:] // cut off "ID="
+
+		case strings.HasPrefix(fields[i], "length="):
+			chrom.Size, err = strconv.Atoi(fields[i][7:]) // cut off "length=" and parse
+			if err != nil {
+				log.Panicf("trouble parsing length field in vcf header.\n Tried to parse '%s'", fields[i][7:])
+			}
+		}
+	}
+
+	// Add chrom to map
+	if _, alreadySawChrom := chroms[chrom.Name]; alreadySawChrom {
+		log.Fatalf("ERROR: contig names in header must be unique. Saw %s multiple times", chrom.Name)
+	}
+	chroms[chrom.Name] = chrom
+	return chroms
```

Comment:
Are you passing back the has because you have modified it (added to it)?  Is this needed?  If it were a slice, yes.  However, I am not sure if this is true of a map.

Reply:
Good point. Removed.